openmeter/billing/service/lineservice/usagebasedline_test.go

```diff
@@ -0,0 +1,1366 @@
+		})
+	})
+
+	// Min spend is always billed in arrears => we are billing it for the last line
+	t.Run("no usage, last line in period, min spend set", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.UnitPrice{
+				Amount:        alpacadecimal.NewFromFloat(10),
+				MinimumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(100)),
+			}),
+			lineMode: lastInPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(0),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: minimum spend",
+					Amount:                 alpacadecimal.NewFromFloat(100),
+					Quantity:               alpacadecimal.NewFromFloat(1),
+					ChildUniqueReferenceID: UnitPriceMinSpendChildUniqueReferenceID,
+					Period:                 &ubpTestFullPeriod,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+				},
+			},
+		})
+	})
+
+	// Usage is billed regardless of line position
+	t.Run("usage present", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.UnitPrice{
+				Amount: alpacadecimal.NewFromFloat(100),
+			}),
+			lineMode: singlePerPeriodLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(10),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: usage in period",
+					Amount:                 alpacadecimal.NewFromFloat(100),
+					Quantity:               alpacadecimal.NewFromFloat(10),
+					ChildUniqueReferenceID: UnitPriceUsageChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+				},
+			},
+		})
+	})
+
+	t.Run("usage present, mid line", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.UnitPrice{
+				Amount: alpacadecimal.NewFromFloat(100),
+			}),
+			lineMode: midPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(10),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: usage in period",
+					Amount:                 alpacadecimal.NewFromFloat(100),
+					Quantity:               alpacadecimal.NewFromFloat(10),
+					ChildUniqueReferenceID: UnitPriceUsageChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+				},
+			},
+		})
+	})
+
+	// Max spend is always honored
+	t.Run("usage present, max spend set, but not hit", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.UnitPrice{
+				Amount:        alpacadecimal.NewFromFloat(10),
+				MaximumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(100)),
+			}),
+			lineMode: midPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(10),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: usage in period",
+					Amount:                 alpacadecimal.NewFromFloat(10),
```

Comment:
shouldn't this be 100?

Reply:
No, there was 10 usage, at 10/unit price. The amount here reflects the unit price. We can rename it perUnitAmount. But on a line level this is always per unit amount.